src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java

```diff
@@ -97,11 +100,21 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
              */
 
             Path indexPath = PathUtils.get(directory, hnswFiles.get(0));
-            final KNNIndex index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName());
-            final KNNQueryResult[] results = index.queryIndex(
-                    knnQuery.getQueryVector(),
-                    knnQuery.getK()
-            );
+            final KNNQueryResult[] results;
+
+            if (fieldAttributes.containsValue(NmsLibVersion.V206.getBuildVersion())) {
```

Comment:
How about we pull out the methods in KNNFIndex and KNNIndex to interface so that we could refer with one object to keep it more cleaner? Something like

```
KNNIndex index;
if (fieldAttributes.containsValue(NmsLibVersion.V206.getBuildVersion())) {
       index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName());
} else {
       index = knnIndexCache.getFIndex(indexPath.toString(), knnQuery.getIndexName());
}
results = index.queryIndex(knnQuery.getQueryVector(), knnQuery.getK());



```

Reply:
this is may make the log of changes in KNNIndex. In KNNIndex class, it would load the specific library, if we want to pull out the methods `getIndex`, we need an abstraction of KNNIndex and verify the engine we are using.
Let me think about it how to do it.

Reply:
> How about we pull out the methods in KNNFIndex and KNNIndex to interface so that we could refer with one object to keep it more cleaner? Something like

in the latest commits, I used KNNIndex as abstract class. `KNNFaissIndex` and `KNNNmsLibIndex` extends the KNNINdex. 
`KNNIndexCacheEntry` only save the KNNIndex.

if you have any other idea, pls let me know